src/main/java/io/cryostat/discovery/DiscoveryStorage.java

```diff
@@ -0,0 +1,234 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package io.cryostat.discovery;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.UUID;
+
+import javax.inject.Provider;
+
+import io.cryostat.core.log.Logger;
+import io.cryostat.core.net.discovery.JvmDiscoveryClient.EventKind;
+import io.cryostat.platform.AbstractPlatformClient;
+import io.cryostat.platform.ServiceRef;
+import io.cryostat.platform.discovery.AbstractNode;
+import io.cryostat.platform.discovery.BaseNodeType;
+import io.cryostat.platform.discovery.EnvironmentNode;
+import io.cryostat.platform.discovery.TargetNode;
+
+@Deprecated
+/**
+ * @deprecated TODO remove this, it's a temporary stub for a database
+ */
+public class DiscoveryStorage extends AbstractPlatformClient {
+
+    private final Provider<UUID> uuid;
+    private final Map<UUID, PluginInfo> map = new HashMap<>();
+    private final Logger logger;
+
+    DiscoveryStorage(Provider<UUID> uuid, Logger logger) {
+        this.uuid = uuid;
+        this.logger = logger;
+    }
+
+    @Override
+    public void start() throws IOException {
+        // TODO persist plugin infos (with empty subtrees) to disk on shutdown, and reinitialize map
+        // here. Then, perform POST on each callback URI to check it's still there and prompt it to
+        // update us with its subtree
+    }
+
+    public UUID register(String realm, URI callback) throws RegistrationException {
+        if (map.values().stream().map(PluginInfo::getRealm).anyMatch(realm::equals)) {
+            throw new RegistrationException(realm);
+        }
+        EnvironmentNode subtree = new EnvironmentNode(realm, BaseNodeType.REALM);
+        UUID nextId = uuid.get();
+        map.put(nextId, new PluginInfo(realm, callback, subtree));
+        return nextId;
+    }
+
+    public Set<AbstractNode> update(UUID id, Set<AbstractNode> children) {
+        try {
+            validateId(id);
+            EnvironmentNode previousTree = map.get(id).getSubtree();
+
+            PluginInfo updatedInfo = new PluginInfo(map.get(id), children);
+            map.put(id, updatedInfo);
+
+            EnvironmentNode currentTree = updatedInfo.getSubtree();
+
+            Set<TargetNode> previousLeaves = findLeavesFrom(previousTree);
+            Set<TargetNode> currentLeaves = findLeavesFrom(currentTree);
+
+            Set<TargetNode> added = new HashSet<>(currentLeaves);
+            added.removeAll(previousLeaves);
+
+            Set<TargetNode> removed = new HashSet<>(previousLeaves);
+            removed.removeAll(currentLeaves);
+
```

Comment:
It looks like to me that the `currentLeaves` set will always be bigger than the `previousLeaves` set because the `currentTree`'s children are instantiated from the previous tree's children and the `children` parameter of `update` are always added to the `currentTree.subtree`.

Reply:
```
previous = [0, 1, 2]
current = [1, 2]

added = copy([1, 2])
added = added - [0, 1, 2] // [ ] empty set

removed = copy([0, 1, 2])
removed = removed - [1, 2] // [ 0 ]
```

```
previous = [0, 1]
current = [0, 1, 2]

added = copy([0, 1, 2])
added = added - [0, 1] // [ 2 ]

removed = copy([0, 1)
removed = removed - [0, 1, 2] // [ ] empty set
```

> because the currentTree's children are instantiated from the previous tree's children and the children parameter of update are always added to the currentTree.subtree.

`currentTree` is the updated tree given to us by the plugin service, but it is not "instantiated from the previous tree's children" - the previous subtree root's name and nodeType are copied, but the children are only the ones given by the update. The algorithm here is detecting what elements were added/removed and using that to emit LOST/FOUND discovery notifications.


```java
            PluginInfo updatedInfo = new PluginInfo(map.get(id), children);
```

I think this is the potentially confusing line, since this looks like a plain copy constructor. The first argument is the `PluginInfo` in storage from the last update (or from registration time), from which the name and nodeType are copied. The second argument is the set of children from the current update.

Reply:
Yeah, I was getting confused at the constructor part
```
            this.subtree =
                    new EnvironmentNode(original.subtree.getName(), original.subtree.getNodeType());
            subtree.addChildren(children);
```
I thought `subtree.addChildren()` was adding to existing children somehow, but I understand now thanks!